tchannel-core/src/main/java/com/uber/tchannel/handlers/LoadControlHandler.java

```diff
@@ -0,0 +1,77 @@
+package com.uber.tchannel.handlers;
+
+import io.netty.channel.ChannelConfig;
+import io.netty.channel.ChannelDuplexHandler;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.ChannelPromise;
+import java.util.Objects;
+import java.util.concurrent.atomic.AtomicIntegerFieldUpdater;
+import org.jetbrains.annotations.NotNull;
+
+/**
+ * Provides backpressure for a Netty pipeline.
+ *
+ * Keeps track of (#reads - #writes) as {@link #outstanding}.
+ *
+ * When the number of outstanding requests surpasses the {@link #high} water mark, reading is paused.
+ *
+ * When the number of outstanding requests goes blow the {@link #low} water mark, reading is resumed.
+ *
+ * This signals the upstream producer (the client) to back off / slow down, because it will not be able to write any
+ * more requests to the connection.
+ *
+ * Note: STATEFUL HANDLER (use new instance per pipeline)
+ */
+public final class LoadControlHandler extends ChannelDuplexHandler {
+
+    private static final AtomicIntegerFieldUpdater<LoadControlHandler> OUTSTANDING_UPDATER =
```

Comment:
do we really need this to be atomic? It is definitely performance hit and I guess that we don't really care about the exact precision of **on/off**  switch.

Reply:
Verified that this does not need to be atomic after Netty 4: https://stackoverflow.com/questions/46508433/concurrency-in-netty